### PR TITLE
Use the new `DeviceOrd` in the linalg module.

### PR DESCRIPTION
**Pull request by trivialfis**

### CODE REVIEW

File: src/tree/updater_gpu_hist.cu [hunk 1]
```diff
@@ -511,8 +511,8 @@ struct GPUHistMakerDevice {
     }
 
     CHECK(p_tree);
-    dh::safe_cuda(cudaSetDevice(ctx_->gpu_id));
-    CHECK_EQ(out_preds_d.DeviceIdx(), ctx_->gpu_id);
+    dh::safe_cuda(cudaSetDevice(ctx_->Ordinal()));
+    CHECK(out_preds_d.Device().IsCUDA());
```
Comment (hcho3) [inline src/tree/updater_gpu_hist.cu:515]: Shouldn't this line compare the device of `out_preds_d` with `ctx_->Device()`?

**Review by hcho3 — APPROVED** [review]

**Comment by hcho3** [comment]
Nice job abstracting the `gpu_id`. I have one question.

File: src/tree/updater_gpu_hist.cu (hunk 1 above)
Comment (trivialfis) [reply src/tree/updater_gpu_hist.cu:515]: It's just CUDA, we won't be using GPU hist if context is CPU.

Comment (hcho3) [reply src/tree/updater_gpu_hist.cu:515]: What if GPU id mismatches?

Comment (trivialfis) [reply src/tree/updater_gpu_hist.cu:515]: That's a bug and might segfault due to accessing the wrong memory

Comment (hcho3) [reply src/tree/updater_gpu_hist.cu:515]: Yes, so shouldn't we have `CHECK_EQ(out_preds.Device(), ctx_->Device())` ?

Comment (trivialfis) [reply src/tree/updater_gpu_hist.cu:515]: I will have an additional check for matching context.

Comment (trivialfis) [reply src/tree/updater_gpu_hist.cu:514]: Done.

**Pull request merged**